File: src/port.rs
use std::collections::HashMap;

use crate::model::ProxyNode;

/// Assigns deterministic ports to nodes, avoiding collisions with
/// ports already in the global `existing_map` (which includes all groups).
/// Returns only the new/updated entries for this batch of nodes.
pub fn assign_ports(
    nodes: &mut [ProxyNode],
    range_start: u16,
    existing_map: &HashMap<String, u16>,
) -> HashMap<String, u16> {
    let mut new_entries = HashMap::new();

    // Collect all occupied ports (from other groups that are already assigned)
    let occupied: std::collections::HashSet<u16> = existing_map.values().copied().collect();
    let mut next_port = range_start;

    for node in nodes.iter_mut() {
        // Reuse existing port if this node already has one
        if let Some(&port) = existing_map.get(&node.name) {
            node.assigned_port = port;
            new_entries.insert(node.name.clone(), port);
            continue;
        }

        // Find the next available port not in occupied set or already assigned
        while occupied.contains(&next_port) || new_entries.values().any(|&p| p == next_port) {
            next_port += 1;
        }

        node.assigned_port = next_port;
        new_entries.insert(node.name.clone(), next_port);
        next_port += 1;
    }

    new_entries
}

File: zig/src/port.zig
const std = @import("std");
const model = @import("model.zig");

pub fn assignPorts(
    gpa: std.mem.Allocator,
    nodes: []model.ProxyNode,
    range_start: u16,
    existing_map: *std.StringHashMap(u16),
) !std.StringHashMap(u16) {
    var new_entries = std.StringHashMap(u16).init(gpa);
    errdefer new_entries.deinit();

    // Collect occupied ports
    var occupied = std.AutoHashMap(u16, void).init(gpa);
    defer occupied.deinit();
    var it = existing_map.valueIterator();
    while (it.next()) |p| {
        try occupied.put(p.*, {});
    }

    var next_port: u16 = range_start;
    for (nodes) |*node| {
        if (existing_map.get(node.name)) |port| {
            node.assigned_port = port;
            try new_entries.put(try gpa.dupe(u8, node.name), port);
            continue;
        }

        while (occupied.contains(next_port)) {
            next_port +|= 1;
            if (next_port == 0) return error.PortExhausted;
        }

        node.assigned_port = next_port;
        try new_entries.put(try gpa.dupe(u8, node.name), next_port);
        try occupied.put(next_port, {});
        next_port +|= 1;
    }

    return new_entries;
}
